File: cmd/kavach/internal/commands/gates/posttool.go
// Package gates provides hook gates for Claude Code.
// posttool.go: Post-tool umbrella gate (PostToolUse for non-write tools).
// Routes by tool name to 1-2 L3 gates.
package gates

import (
	"fmt"
	"os"

	"github.com/claude/shared/pkg/context"
	"github.com/claude/shared/pkg/dag"
	"github.com/claude/shared/pkg/enforce"
	"github.com/claude/shared/pkg/hook"
	"github.com/claude/shared/pkg/stmlog"
	"github.com/claude/shared/pkg/telemetry"
	"github.com/spf13/cobra"
)

var postToolHookMode bool

var postToolCmd = &cobra.Command{
	Use:   "post-tool",
	Short: "Post-tool umbrella gate (memory|context|research|task)",
	Run:   runPostToolGate,
}

func init() {
	postToolCmd.Flags().BoolVar(&postToolHookMode, "hook", false, "Hook mode")
}

func runPostToolGate(cmd *cobra.Command, args []string) {
	if !postToolHookMode {
		cmd.Help()
		return
	}

	span := telemetry.StartSpan("post-tool")
	defer span.End()

	input := hook.MustReadHookInput()
	span.SetTool(input.ToolName)

	// Only load session for tools that need it
	var session *enforce.SessionState
	needsSession := input.ToolName == "WebSearch" || input.ToolName == "WebFetch" ||
		input.ToolName == "TaskCreate" || input.ToolName == "TaskUpdate" ||
		input.ToolName == "TaskOutput"
	if needsSession {
		session = enforce.GetOrCreateSession()
		span.SetSessionLoaded(true)
	}

	switch input.ToolName {
	case "Bash":
		command := input.GetString("command")
		if command != "" && stmlog.IsBashSignificant(command) {
			stmlog.AppendEvent("", "bash", command, "")
		}
		hook.ExitSilent()

	case "Read":
		filePath := input.GetString("file_path")
		if filePath != "" {
			context.TrackFileRead(filePath)
		}
		hook.ExitSilent()

	case "Glob":
		path := input.GetString("path")
		pattern := input.GetString("pattern")
		if path != "" {
			context.TrackFileRead(path)
		} else if pattern != "" {
			context.TrackFileRead("glob:" + pattern)
		}
		hook.ExitSilent()

	case "Grep":
		path := input.GetString("path")
		pattern := input.GetString("pattern")
		if path != "" {
			context.TrackFileRead(path)
		} else if pattern != "" {
			context.TrackFileRead("grep:" + pattern)
		}
		hook.ExitSilent()

	case "Task":
		agentType := input.GetString("subagent_type")
		desc := input.GetString("description")
		if agentType != "" {
			context.TrackAgentCompletion(agentType)
			stmlog.AppendEvent("", "agent_"+agentType, desc, "")
		}
		hook.ExitSilent()

	case "WebSearch", "WebFetch":
		topic := input.GetString("query") // WebSearch
		if topic == "" {
			topic = input.GetString("url") // WebFetch
		}
		session.MarkResearchDoneWithTopic(topic)
		span.SetResult("research_marked")
		hook.ExitSilent()

	case "TaskCreate":
		postToolTaskCreate(input, session)

	case "TaskUpdate":
		postToolTaskUpdate(input, session)

	case "TaskOutput":
		postToolTaskOutput(input)

	default:
		hook.ExitSilent()
	}
}

// postToolTaskCreate handles post-creation tracking.
func postToolTaskCreate(input *hook.Input, session *enforce.SessionState) {
	subject := input.GetString("subject")
	session.TasksCreated++
	session.SetCurrentTask(subject)
	session.Save()

	// DAG tracking
	if state, err := dag.Load(session.SessionID); err == nil {
		_, _, directive := dag.HandleTaskEvent(state, "TaskCreate", input.ToolInput)
		if err := dag.Save(state); err != nil {
			fmt.Fprintf(os.Stderr, "[TASK_DAG] Save error: %v\n", err)
		}
		if directive != "" {
			hook.ExitModifyTOONWithModule("TASK_CREATE_POST", map[string]string{
				"dag_directive": "active",
			}, directive)
		}
	}

	hook.ExitSilent()
}

// postToolTaskUpdate handles post-update tracking.
func postToolTaskUpdate(input *hook.Input, session *enforce.SessionState) {
	status := input.GetString("status")
	subject := input.GetString("subject")

	if status == "completed" || status == "deleted" {
		session.TasksCompleted++
		session.ClearTask()
	} else if status == "in_progress" && subject != "" {
		session.SetTask(subject, status)
	}
	session.Save()

	// DAG advancement
	if state, err := dag.Load(session.SessionID); err == nil {
		complete, needsAegis, directive := dag.HandleTaskEvent(state, "TaskUpdate", input.ToolInput)
		if err := dag.Save(state); err != nil {
			fmt.Fprintf(os.Stderr, "[TASK_DAG] Save error: %v\n", err)
		}
		if complete && needsAegis {
			hook.ExitModifyTOON("TASK_UPDATE_DAG_COMPLETE", map[string]string{
				"dag_status": "complete",
				"action":     "Run kavach orch aegis for final verification",
			})
		}
		if directive != "" {
			hook.ExitModifyTOONWithModule("TASK_UPDATE_DAG_ADVANCE", map[string]string{
				"dag_status": "advancing",
			}, directive)
		}
	}

	hook.ExitSilent()
}

// postToolTaskOutput handles task output retrieval tracking.
func postToolTaskOutput(input *hook.Input) {
	taskID := input.GetString("task_id")
	if taskID == "" {
		hook.ExitSilent()
	}

	health := GetTaskHealth()
	if issue := health.RecordTaskOutput(taskID, ""); issue != nil {
		hook.ExitModifyTOON("TASK_OUTPUT", map[string]string{
			"task_id":      taskID,
			"health_event": issue.IssueType,
			"health_desc":  issue.Description,
		})
	}

	hook.ExitSilent()
}

File: crates/kavach-cli/src/commands/gates/post_tool.rs
//! Post-tool umbrella gate (PostToolUse for non-write tools).
//! Routes by tool name: memory | context | research | task

use std::collections::HashMap;

use crate::hook::{self, HookInput};
use crate::session;

pub fn run(hook_mode: bool) -> anyhow::Result<()> {
    if !hook_mode {
        crate::commands::cli_print("gates post-tool: use --hook flag");
        return Ok(());
    }

    let input = hook::must_read_hook_input();
    let result = dispatch(&input);

    match result {
        Ok(()) => Ok(()),
        Err(e) if hook::is_hook_exit(&e) => Ok(()),
        Err(e) => Err(e),
    }
}

fn dispatch(input: &HookInput) -> anyhow::Result<()> {
    match input.get_tool_name() {
        "Bash" => post_tool_bash(input),
        "Read" | "Glob" | "Grep" => post_tool_read(input),
        "Task" => post_tool_task(input),
        "WebSearch" | "WebFetch" => post_tool_research(input),
        "TaskCreate" => post_tool_task_create(input),
        "TaskUpdate" => post_tool_task_update(input),
        "TaskOutput" => post_tool_task_output(input),
        _ => hook::exit_silent(),
    }
}

fn post_tool_bash(_input: &HookInput) -> anyhow::Result<()> {
    hook::exit_silent()
}

fn post_tool_read(_input: &HookInput) -> anyhow::Result<()> {
    hook::exit_silent()
}

fn post_tool_task(input: &HookInput) -> anyhow::Result<()> {
    let agent_type = input.get_string("subagent_type");
    if !agent_type.is_empty() {
        let mut kvs = HashMap::new();
        kvs.insert("agent".into(), agent_type);
        kvs.insert("status".into(), "completed".into());
        hook::exit_modify_toon("AGENT_COMPLETE", &mut kvs)?;
    }
    hook::exit_silent()
}

fn post_tool_research(input: &HookInput) -> anyhow::Result<()> {
    let mut session = session::get_or_create_session();

    let topic = if input.get_tool_name() == "WebSearch" {
        input.get_string("query")
    } else {
        input.get_string("url")
    };

    session.mark_research_done_with_topic(&topic);

    hook::exit_silent()
}

fn post_tool_task_create(input: &HookInput) -> anyhow::Result<()> {
    let mut session = session::get_or_create_session();
    let subject = input.get_string("subject");

    session.tasks_created += 1;
    session.set_current_task(&subject);

    hook::exit_silent()
}

fn post_tool_task_update(input: &HookInput) -> anyhow::Result<()> {
    let mut session = session::get_or_create_session();
    let status = input.get_string("status");

    if status == "completed" || status == "deleted" {
        session.tasks_completed += 1;
        session.clear_task();
    } else if status == "in_progress" {
        let subject = input.get_string("subject");
        if !subject.is_empty() {
            session.set_current_task(&subject);
        }
    }

    hook::exit_silent()
}

fn post_tool_task_output(_input: &HookInput) -> anyhow::Result<()> {
    hook::exit_silent()
}
